### ISSUE TITLE
File deletion fails on first attempt through union mount

### ISSUE THREAD

**Issue by fuzxi** [rclone/rclone#5615]
#### The associated forum post URL from `https://forum.rclone.org`

https://forum.rclone.org/t/file-deletion-fails-on-first-attempt-through-union-mount/26235

#### What is the problem you are having with rclone?

My setup is pretty common. I have a Google Drive remote mounted through a crypt remote. I also have a union remote which combines the crypt remote and local disk storage to simplify Radarr/Sonarr file management. My mount commands are as follows.

Crypt remote:
`rclone mount --drive-impersonate=[email redacted] --allow-other --buffer-size 128M --dir-cache-time 168h --drive-chunk-size 64M --log-level INFO --vfs-cache-mode off --vfs-read-chunk-size 128M --vfs-read-chunk-size-limit off gcrypt:crypt /media/gcrypt`

Union remote:
`rclone mount --allow-other media-union: /data/media`

When deleting files on cloud storage through the union remote, rclone will almost always through an "Operation not permitted" error on the first attempt, then succeeds on the second. If I delete the file through the crypt remote, e.g. `/media/gcrypt/media` instead of `/data/media`, I don't get an error. I also do not get this error if I create an empty file and then attempt to delete it, regardless of whether the file was created through the union remote or directly through the crypt remote.

If I use `rm` instead of `rclone delete`, I get the same error, the file disappears on the union mount, but still exists on the crypt mount.

This causes issues for me with Sonarr/Radarr, as when they upgrade old media with higher quality versions, the old media will not be deleted from the crypt mount and will reappear on the union mount after a reboot, so I end up with two copies of the same thing.

Also, I'd like to add that this is not a permission error related to Radarr/Sonarr or Docker. I can run `docker exec -it radarr bash` to get an interactive shell inside the Radarr container and the same behavior occurs as when I attempt to run `rm` outside the container.

I have also tested with `--vfs-cache-mode minimal`. This does not make a difference.

#### What is your rclone version (output from `rclone version`)

1.56.1. This also occurred on versions 1.56.0 and 1.55.0.

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)

64-bit Debian Linux (Buster).

#### Which cloud storage system are you using? (e.g. Google Drive)

Google Drive.

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)

`rclone delete -vv /data/media/movies/Nobody\ \(2021\)/Nobody\ \(2021\)\ Remux-1080p.en.srt`

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)

````
2021/09/22 00:38:32 DEBUG : rclone: Version "v1.56.1" starting with parameters ["rclone" "delete" "-vv" "/data/media/movies/Nobody (2021)/Nobody (2021) Remux-1080p.en.srt"]
2021/09/22 00:38:32 DEBUG : Creating backend with remote "/data/media/movies/Nobody (2021)/Nobody (2021) Remux-1080p.en.srt"
2021/09/22 00:38:32 DEBUG : Using config file from "/home/user/.config/rclone/rclone.conf"
2021/09/22 00:38:32 DEBUG : fs cache: adding new entry for parent of "/data/media/movies/Nobody (2021)/Nobody (2021) Remux-1080p.en.srt", "/data/media/movies/Nobody (2021)"
2021/09/22 00:38:32 DEBUG : Waiting for deletions to finish
2021/09/22 00:38:32 DEBUG : Nobody (2021) Bluray-2160p.en.forced.srt: Excluded
2021/09/22 00:38:32 DEBUG : Nobody (2021) Bluray-2160p.en.srt: Excluded
2021/09/22 00:38:32 DEBUG : Nobody (2021) Bluray-2160p.mkv: Excluded
2021/09/22 00:38:32 ERROR : Nobody (2021) Remux-1080p.en.srt: Couldn't delete: remove /data/media/movies/Nobody (2021)/Nobody (2021) Remux-1080p.en.srt: operation not permitted
2021/09/22 00:38:32 ERROR : Attempt 1/3 failed with 2 errors and: failed to delete 1 files
2021/09/22 00:38:32 DEBUG : Waiting for deletions to finish
2021/09/22 00:38:32 DEBUG : Nobody (2021) Bluray-2160p.en.forced.srt: Excluded
2021/09/22 00:38:32 DEBUG : Nobody (2021) Bluray-2160p.en.srt: Excluded
2021/09/22 00:38:32 DEBUG : Nobody (2021) Bluray-2160p.mkv: Excluded
2021/09/22 00:38:32 ERROR : Attempt 2/3 succeeded
2021/09/22 00:38:32 DEBUG : 2 go routines active
````

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by darthShadow** [comment]
Can you check that your permissions are proper for the local storage i.e. you are able to create and delete files on there via the same user that rclone is running as?

**Comment by fuzxi** [comment]
Yes, I can, on both the union mount and the crypt mount.

**Comment by darthShadow** [comment]
Can you please add the logs from the union mount while deleting a file which fails?

**Comment by fuzxi** [comment]
[Log for failure using rm](https://github.com/rclone/rclone/files/7215124/rclone-union-1.log)
[Log for "success" using rclone delete (file still exists on crypt remote)](https://github.com/rclone/rclone/files/7215125/rclone-union-2.log)

**Comment by darthShadow** [comment]
It seems fine. I am still guessing at a permissions issue somewhere but not sure what exactly since I haven't ever used unionfs myself.

You will have to wait for @ncw (or someone else who has used it before) to chime in either here or on the forums, sorry.

**Comment by ncw** [comment]
Studying those logs it looks like that the remove which gave the error `permission denied` actually worked?

Can you confirm that? Try `rm` the file then `ls` to see if it is still there?

The `permission denied` error came from the local backend when it was looking for the object to delete. Precisely it was when rclone ran `lstat` on the object path, it got a permissions error.

So it is really likely that there is some permissions problem with the local directory. Maybe rclone mount is running as a different user or something like that?

**Comment by fuzxi** [comment]
That's the problem, files are successfully deleted from the union mount but still exist on the crypt remote. 
e.g. if I run `rm /data/media/movies/Nobody\ \(2021\)/Nobody\ \(2021\)\ Remux-1080p.en.srt`, the file will no longer exist on `/data/media/movies` but will still exist on `/media/gcrypt/media/movies`, and will show up in `/data/media/movies` again after remounting the union remote. However, I can write to the union mount and it shows up and behaves as expected on the local storage half of the union. It almost feels like the crypt remote is read-only, but I can write to the crypt mount directly.

Both mounts are handled by systemd and are run as the same user that ran these test commands. I just stopped both systemd units, unmounted them, and remounted by running the mount commands directly and there was no change, so the systemd units aren't related. Both of the test files I've posted logs for are also owned by the same user, with 0644 permissions.

The union section of my `rclone.conf` is:

```
[media-union]
type = union
remotes = gcrypt:crypt/media /media/3TB/media
```

I just tried changing the order of the remotes, i.e. to `remotes = /media/3TB/media gcrypt:crypt/media`, and it runs "in reverse", as expected. Writing to the union mount will write to the crypt remote instead of local storage. But the `Operation not permitted` error is reversed as well! With this configuration, attempting to delete a file on local storage through the union mount gives the error, but everything works as expected on the crypt remote half of the union.


**tl;dr** it looks like the issue is that the second remote in the union is mounted read-only, even for deletions ("action functions" in the union remote docs). I just tried mounting the union remote with `rclone mount --allow-other --union-action-policy epall media-union: /data/media` in case the default action policy got changed somehow, but this didn't make any difference.

**Comment by amidevous** [comment]
try

[local]
type = local
[media-union]
type = union
remotes = gcrypt:crypt/media local:media/3TB/media

**Comment by fuzxi** [comment]
Please read the issue before posting. We have already ruled out incorrect permissions. The error in question is thrown for whichever remote or local directory is listed last in the union mount. It is not specific to the local drive.

**Comment by ncw** [comment]
@fuxi can you try to make a self contained reproducer for this please?

Here is my attempt

```
mkdir /tmp/a /tmp/b
touch /tmp/a/IN-A
touch /tmp/b/IN-B
rclone mount -vv ':union,upstreams="/tmp/a /tmp/b":' /mnt/tmp/
```

then

```
rclone delete -vv /mnt/tmp/IN-A
# success
rclone delete -vv /mnt/tmp/IN-B
# success
```

That didn't reproduce the problem though...

**Comment by fuzxi** [comment]
I have my hands full now. I won't be able to touch this for a while.

**Comment by ncw** [comment]
@fuzxi no problems. Just put the reproducer in the issue when you are ready and I'll pick it up.
